lib: os: Rename icmsg_buf to spsc_pbuf

nordic-krch opened the pull request:
Move icmsg_buf to lib/os and rename to spsc_pbuf (Single Producer
Single Consumer Packet Buffer). It is a generic module and initially
was created as internal module for ipc service.

nordic-krch committed:
lib: os: Rename icmsg_buf to spsc_pbuf

Move icmsg_buf to lib/os and rename to spsc_pbuf (Single Producer
Single Consumer Packet Buffer). It is a generic module and initially
was created as internal module for ipc service.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: kernel: Add documentation to spsc_pbuf

Add documentation to spsc_pbuf module.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@emob-nordic @pdunaj I would like to use this buffer elsewhere and soon extend it as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/43544#issuecomment-1070755803

carlocaione commented:
Can you rename also the IPC service backend?

nordic-krch committed:
lib: os: spsc_pbuf: Add option to use cache

Add flags option to init call and a flag to use cache.
Add Kconfig choice to pick how to approach cache. Cache can be
enforced in all spsc_pbuf instances, disable in all, or runtime selected
based on configuration flag. Option is added to allow memory footprint
savings.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

emob-nordic commented on tests/lib/spsc_pbuf/src/main.c:
```diff
@@ -5,45 +5,45 @@
 	static uint8_t rbuf[198];
 	static uint8_t message[20] = {'a'};
-	struct icmsg_buf *ib;
+	struct spsc_pbuf *ib;
```
Please also change ib -> pbuf (packet buffer) or similar. All across the file and in similar places.

nordic-krch replied:
done, changed to `pb`

eobalski reviewed (changes requested):
The short name is a bit complicated, but i don't see any better suggestion. Maybe just circular packet buffer instead (`cpb`)?

I nacked because of small fixes commented inline.

nordic-krch commented:
regarding renaming of ipc service. I wonder if it should be renamed to something that would indicate its feature (lightweight, simplicity). Was thinking about something like `ipc_lite`? That will tell more to the new user than `ipc_spsc_pbuf` which tells nothing.

nordic-krch commented:
^ @pdunaj @emob-nordic @carlocaione

carlocaione commented:
`ipc_lite` is way too generic IMO (what if in the future we add a new backend even lighter?). If the the service is actually using `spsc_pbuf` behind the curtains I this that `ipc_spsc_pbuf` is a proper name for that. My 2 cents.

carlocaione commented:
> What's the problem with leaving this `icmsg`?

Because especially in this case the IPC service is really a wrapper around `spsc_pbuf` so why using two different names? It's just confusing for the user.

nordic-krch committed:
tests: lib: spsc_pbuf: Exclude ast1030_evb from test

This board has external cache enabled but fails to link when cache
API is used. Temporary disabling.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -22,7 +22,7 @@ config IPC_SERVICE_BACKEND_ICMSG
 	bool "ICMSG backend with ICMsg buffer"
```
done

nordic-krch commented:
I'm a bit reluctant to change the name of the backend as there is no consensus.

dcpleung commented on doc/kernel/data_structures/spsc_pbuf.rst:
```diff
@@ -0,0 +1,17 @@
+is a single context that produces packages and a single context that consumes the
+data.
+
+Implementation is focused on perfomance and memory footprint.
```
s/perfomance/performance/ (missing 'r')

dcpleung commented on doc/kernel/data_structures/spsc_pbuf.rst:
```diff
@@ -0,0 +1,17 @@
+A :dfn:`Single Producer Single Consumer Packet Buffer (SPSC_PBUF)` is a circular
+buffer, whose contents are stored in first-in-first-out order. Variable size
+packets are stored in the buffer. Packet buffer works under assumption that there
+is a single context that produces packages and a single context that consumes the
```
Maybe `packets` instead of `packages`?

pdunaj commented:
> I'm a bit reluctant to change the name of the backend as there is no consensus.

I am personally against it. The current name is as good as any other and the backend is already recognized by its name in various places. Name change will lead to additional maintenance work giving absolutely no benefits.

eobalski reviewed (changes requested):
Please change to `project(spsc_pbuf)` https://github.com/zephyrproject-rtos/zephyr/blob/784ed03bb061788927bd53699f376f0f41d13e1a/tests/lib/spsc_pbuf/CMakeLists.txt#L5

Other than than, its ok.

nordic-krch commented:
@carlocaione can we skip renaming of ipc backend? I would like to push it forward.

carlocaione commented:
> @carlocaione can we skip renaming of ipc backend? I would like to push it forward.

yup, no problem.

nordic-krch commented:
@dcpleung @pdunaj can you reapprove. No changes, only upmerge.
@emob-nordic can you take a final look. Your last comment got applied.
